lib/auth/password.go

```diff
@@ -416,47 +435,112 @@ func (s *Server) changeUserSecondFactor(req ChangePasswordWithTokenRequest, toke
 		if secondFactor == constants.SecondFactorU2F {
 			return trace.BadParameter("user %q sent an OTP token during password reset but cluster only allows U2F for second factor", username)
 		}
-		secrets, err := s.Identity.GetUserTokenSecrets(ctx, req.TokenID)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-
-		dev, err := services.NewTOTPDevice("otp", secrets.GetOTPKey(), s.clock.Now())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.checkTOTP(ctx, username, req.SecondFactorToken, dev); err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.UpsertMFADevice(ctx, username, dev); err != nil {
+		if _, err := s.createNewTOTPDevice(ctx, newTOTPDeviceRequest{
+			tokenID:           req.GetTokenID(),
+			username:          username,
+			deviceName:        req.GetDeviceName(),
+			secondFactorToken: req.GetSecondFactorToken(),
+		}); err != nil {
 			return trace.Wrap(err)
 		}
 
 		return nil
 	}
+
 	if req.U2FRegisterResponse != nil {
 		if secondFactor == constants.SecondFactorOTP {
 			return trace.BadParameter("user %q sent a U2F registration during password reset but cluster only allows OTP for second factor", username)
 		}
+
 		cfg, err := cap.GetU2F()
 		if err != nil {
 			return trace.Wrap(err)
 		}
 
-		_, err = u2f.RegisterVerify(ctx, u2f.RegisterVerifyParams{
-			DevName:                "u2f",
-			ChallengeStorageKey:    req.TokenID,
-			RegistrationStorageKey: username,
-			Resp:                   *req.U2FRegisterResponse,
-			Storage:                s.Identity,
-			Clock:                  s.GetClock(),
-			AttestationCAs:         cfg.DeviceAttestationCAs,
-		})
-		return trace.Wrap(err)
+		if _, err := s.createNewU2FDevice(ctx, newU2FDeviceRequest{
+			tokenID:    req.GetTokenID(),
+			username:   username,
+			deviceName: req.GetDeviceName(),
+			u2fRegisterResponse: u2f.RegisterChallengeResponse{
+				RegistrationData: req.GetU2FRegisterResponse().GetRegistrationData(),
+				ClientData:       req.GetU2FRegisterResponse().GetClientData(),
+			},
+			cfg: cfg,
+		}); err != nil {
+			return trace.Wrap(err)
+		}
+
+		return nil
 	}
 
 	if secondFactor != constants.SecondFactorOptional {
 		return trace.BadParameter("no second factor sent during user %q password reset", username)
 	}
 	return nil
 }
+
+type newTOTPDeviceRequest struct {
+	tokenID           string
+	username          string
+	deviceName        string
+	secondFactorToken string
+}
+
+func (s *Server) createNewTOTPDevice(ctx context.Context, req newTOTPDeviceRequest) (*types.MFADevice, error) {
+	secrets, err := s.Identity.GetUserTokenSecrets(ctx, req.tokenID)
```

Comment:
This query will always hit the backend. Is that desirable? Consider adding a wrapper for `GetUserTokenSecrets` (and possibly other getters you added to `IdentityService`) to `auth/auth.go`:
```go
func (a *Server) GetUserTokenSecrets(ctx context.Context, tokenID string) (types.UserTokenSecrets, error) {
	return a.GetCache().GetUserTokenSecrets(ctx, tokenID)
}
```
and then calling
```suggestion
	secrets, err := s.GetUserTokenSecrets(ctx, req.tokenID)
```
here.

Reply:
I think it depends on how quickly the cache gets updated (which i'm not sure of)

`GetUserTokenSecrets` in a happy path, is really only a one time call/use. The getter could get triggered more than once from some server error or user error (providing wrong totp token value), but this secret can also change if a user triggers `RotateUserTokenSecrets` (from like a page refresh) which upserts new secrets per call. `RotateUserTokenSecrets` gets called upon user click on a link, which then user can right away enter auth creds that trigger `GetUserTokenSecrets`.

And other getters I defined are probably not good candidates for cache either. Like `recoveryCodes` is probably going to be infrequently read, so I don't think it should be cached?  (we don't cache user password hashes and passwords will be used more often). Other getters like get `recoveryAttempts` and `getMFADevice` also relies on how quickly the cache gets updated.